ethexe/contracts/src/MirrorImpl.sol

```diff
@@ -1,122 +1,56 @@
-        _;
-    }
-
-    /// @dev Non-zero value must be transferred from source to router in functions marked with this modifier.
-    modifier retrievingValue(uint128 value) {
-        if (value != 0) {
-            address routerAddr = router();
-            bool success = _wvara(routerAddr).transferFrom(_source(), routerAddr, value);
-            require(success, "failed to transfer non-zero amount of WVara from source to router");
-        }
-        _;
-    }
-
-    // TODO (breathx): terminated programs compute threshold must always be treated as balance-enough.
-    /// @dev Functions marked with this modifier can be called only after the program is terminated.
-    modifier whenTerminated() {
-        require(inheritor != address(0), "program is not terminated");
-        _;
-    }
-
-    /// @dev Functions marked with this modifier can be called only after the initializer has created the init message.
-    modifier whenInitMessageCreated() {
-        require(nonce > 0, "initializer hasn't created init message yet");
-        _;
-    }
-
-    /// @dev Functions marked with this modifier can be called only after the initializer has created the init message or from the initializer (first access).
-    modifier whenInitMessageCreatedOrFromInitializer() {
-        require(
-            nonce > 0 || _source() == initializer,
-            "initializer hasn't created init message yet; and source is not initializer"
-        );
-        _;
-    }
-
-    /// @dev Functions marked with this modifier can be called only if the program is active.
-    modifier whileActive() {
-        require(inheritor == address(0), "program is terminated");
-        _;
-    }
-
-    /* Operational functions */
-
-    function router() public view returns (address) {
-        return IMirrorProxy(address(this)).router();
+    function initialize(address, address, address, address) public pure {
+        revert("must be called on MirrorAbi contract");
     }
 
     /* Primary Gear logic */
 
-    function sendMessage(bytes calldata _payload, uint128 _value)
-        external
-        whileActive
-        whenInitMessageCreatedOrFromInitializer
-        retrievingValue(_value)
-        returns (bytes32)
-    {
+    function sendMessage(bytes calldata _payload, uint128 _value) external returns (bytes32) {
         bytes32 id = keccak256(abi.encodePacked(address(this), nonce++));
 
         emit MessageQueueingRequested(id, _source(), _payload, _value);
 
         return id;
     }
 
-    function sendReply(bytes32 _repliedTo, bytes calldata _payload, uint128 _value)
-        external
-        whileActive
-        whenInitMessageCreated
-        retrievingValue(_value)
-    {
+    function sendReply(bytes32 _repliedTo, bytes calldata _payload, uint128 _value) external {
         emit ReplyQueueingRequested(_repliedTo, _source(), _payload, _value);
     }
 
-    function claimValue(bytes32 _claimedId) external whenInitMessageCreated {
```

Comment:
I disagree with removing modifiers like `whenInitMessageCreated` from `MirrorImpl` and moving them to `MirrorAbi` for the following reasons:
1. it increases size of `MirrorAbi` bytecode and makes `createProgram` more expensive (since we are making an `extcodecopy` of `MirrorAbi` bytecode in `createProgram`)
2. it generally looks like an anti-pattern to call something `MirrorImpl` but do the actual checks in proxy such as `MirrorAbi`